design/merge-refactor.md

```diff
@@ -0,0 +1,26 @@
+# Design: Refactor the Merge Operation
+
+## Problem Description
+Currently, as part of our design:
+* Apply is used for both commit and merge
+* Merge iterator is implemented using the Diff iterator
+
+Our code for the merge operation got to a level where it's too complex and it is very hard to follow when inserting changes.
+
+## Goals
+* Simplify the merge implementation to make it easier to insert new changes
+* Create code structure that will allow optimization of the merge operation
+
+## Proposed Design
+![merge refactor design](diagrams/merge-refactor-design.png)
+
+
+### Separate between Diff and Merge:
+Currently, the merge operation uses the diff iterator and patches it with the merge iterator.
+Instead of using the diff iterator that holds source iterator and destination iterator, we can use a compare iterator that will get three iterators: source, destination and base. That way, we can optimize the code by comparing the 3 iterators together and skipping ranges by need.
+
+### Separate between Apply and Merge:
```

Comment:
Do we need an Apply? from the description of this section it looks like both commit and merge needs a Writer. Make the actual implementation is written into apply as one that processes incoming Iterator. But it looks like both cases write a commit or metarange writer to serialize changes.

Reply:
Me too!  Some history: Originally Apply was _just_ for the commit operation.  Then we changed it to support merging because we thought that way we would have less code and it would be simpler.  But I think that the opposite occurred.

It is actually harder to write code for an iterator than it is to write code for a writer.  That's because iterator code has to remember where it was at every step, and it must also handle various distractions (e.g. entering a new range and discovering that it is empty).  In contrast, a writer is just a single loop (E.g. it can use just variables rather than instance variables, and it can use many consecutive loops in the code, and if it enters a new range and discovers that it is empty, it can just write out an empty range or skip that write...).

My ideal refactor would probably be:

* Separate "Apply" and "Merge" operations (maybe "Apply" :arrow_right: "Commit").
* Apply and Merge receive Iterator inputs and a metarange writer (in particular, there is *no* iterator over the commit or merge results).
   * Apply receives _two_ iterators: the original values and the diffs.
   * Merge receives _three_ iterators: from, to, and base.